include/LightGBM/dataset.h

```diff
@@ -674,6 +689,43 @@ class Dataset {
 
   void AddFeaturesFrom(Dataset* other);
 
+  /*! \brief Get has_raw_ */
+  inline double has_raw() const { return has_raw_; }
+
+  /*! \brief Set has_raw_ */
+  inline void SetRaw(bool has_raw) { has_raw_ = has_raw; }
+
+  /*! \brief Get size of raw data */
+  inline data_size_t get_raw_size() const {
+    if (raw_data_.size() == 0) {
+      return 0;
+    } else {
+      return raw_data_[0].size();
+    }
+  }
+
+  /*! \brief Resize raw_data_, use current number of featurse */
+  inline void ResizeRaw(int num_rows) {
+    if (static_cast<int>(raw_data_.size()) > num_numeric_features_) {
+      raw_data_.resize(num_numeric_features_);
+    }
+    for (size_t i = 0; i < raw_data_.size(); ++i) {
+      raw_data_[i].resize(num_rows);
+    }
+    int curr_size = raw_data_.size();
+    for (int i = curr_size; i < num_features_; ++i) {
+      int feat_num = numeric_feature_map_[i];
+      if (feat_num > -1) {
+        raw_data_.push_back(std::vector<float>(num_rows, 0));
```

Comment:
so, categorical features is not supported? maybe we should document it.

Reply:
Added some more documentation in `config.h`. Categorical features are still used in splits as normal, but the linear models at each leaf only use numerical features.